examples/src/main/kotlin/io/ethers/examples/tokenswapwitheventlistening/TokenSwapWithEventListening.kt

```diff
@@ -0,0 +1,117 @@
+import io.ethers.providers.WsClient
+import io.ethers.signers.PrivateKeySigner
+import kotlinx.cli.ArgParser
+import kotlinx.cli.ArgType
+import kotlinx.cli.default
+import kotlinx.cli.required
+import java.lang.Exception
+import java.math.BigInteger
+
+/**
+ * Uniswap V2 ETH to ERC20 swap with listening to UniswapV2Pair Sync event
+ */
+class TokenSwapWithEventListening(
+    privateKey: String,
+    routerAddress: String,
+    wsRpcUrl: String,
+    httpRpcUrl: String,
+    private val wethAddress: String,
+    private val tokenAddress: String,
+    private val ethAmount: BigInteger,
+
+    ) {
+    // Init provider for swap and subscription
+    private val wsClient = WsClient(wsRpcUrl)
+    private val wsProvider = Provider(wsClient)
+    private val httpClient = HttpClient(httpRpcUrl)
+    private val httpProvider = Provider(httpClient)
+
+    // Init wallet
+    private val signer = PrivateKeySigner(privateKey)
+
+    private val router = UniswapV2Router02(httpProvider, Address(routerAddress))
+    fun run() {
+        // Get pool address via factory
+        val factoryAddress = router.factory().call(BlockId.LATEST).sendAwait().resultOrThrow()
+        val pairAddress = UniswapV2Factory(wsProvider, factoryAddress)
+            .getPair(Address(wethAddress), Address(tokenAddress))
+            .call(BlockId.LATEST)
+            .sendAwait()
+            .resultOrThrow()
+
+        if (pairAddress == Address.ZERO) {
+            throw Exception("Pair does not exist!")
+        }
+
+        // # Event listening
+        // We catch UniswapV2Pair Sync event when anyone triggers it on observed pairAddress
+        // It is also possible to filter based on topics and block numbers
+        println("Listening for Sync events on pair: $pairAddress")
+        val stream = UniswapV2Pair.Sync
+            .filter(wsProvider)
+            .address(pairAddress)
+            .subscribe()
+            .sendAwait()
+            .resultOrThrow()
+
+        // We use forEachAsync which listens to events in a separate thread, to avoid blocking the caller
+        stream.forEachAsync {
+            println("Found tx (${it.transactionHash}) with Sync event")
+        }
+
+        // # Executing swap that emits Sync event
+        println("Executing swapExactETHForTokens...")
+        val deadline = ((System.currentTimeMillis() / 1000) + 1800).toBigInteger()
+        val call = router.swapExactETHForTokens(
+            BigInteger.ZERO, // amountOutMin
+            arrayOf(Address(wethAddress), Address(tokenAddress)), // path
+            signer.address, // to
+            deadline,
+        )
+        val pendingTx = call.value(ethAmount).send(signer).sendAwait().resultOrThrow()
+
+        println("Wait for transaction: ${pendingTx.hash} to be included in a block...")
+        val receipt = pendingTx.awaitInclusion(retries = 10).resultOrThrow()
+        println("Buy tx ${receipt.transactionHash} was included in block ${receipt.blockNumber}")
```

Comment:
I suggest you add a 5 sec sleep after this line, in case the WS stream is lagging behind so it prints the Sync event.

Reply:
I removed the closing of wsClient to continue listening to events, so our Sync event is always printed.